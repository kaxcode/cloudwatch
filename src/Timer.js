import React, { Component } from 'react';
import { Row, Button, Card, Col } from 'react-materialize';

export default class Timer extends Component {

  state = {
    timer: null,
    counter: 0
  };

  startTimer = () => {
    clearInterval(this.state.timer);
    this.setState({
      counter: 0
    });
    let timer = setInterval(this.tick, 1000);
    this.setState({ timer });
  }

  pauseTimer = () => {
    clearInterval(this.state.timer);
  }

  clearTimer = () => {
    clearInterval(this.state.timer);
    this.setState({ counter: 0 });
  }

  tick = () => {
    this.setState({
      counter: this.state.counter + 1
    });
  }

  render() {
    return (
      <Row>
        <Col s={12} m={3} offset="m2 l4">
          <Card className="white parent-container" textClassName="black-text">
            <h3>CloudWatch</h3>
            <h3>{this.state.counter}</h3>
            <div>
              <Button waves="light" className="green" onClick={this.startTimer}>
                Start
              </Button>
              <Button waves="light" className="red" onClick={this.pauseTimer}>
                Pause
              </Button>
              <Button
                waves="light"
                className="yellow darken-3"
                onClick={this.clearTimer}
              >
                Clear
              </Button>
            </div>
          </Card>
        </Col>
      </Row>
    );
  }
}
