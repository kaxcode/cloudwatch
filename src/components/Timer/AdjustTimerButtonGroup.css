.AdjustTimerButtonGroup__PlusBtn {
  margin-right: 20px;
  margin-left: 20px;
}

.AdjustTimerButtonGroup {
  grid-area: adjustBtns;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
}
