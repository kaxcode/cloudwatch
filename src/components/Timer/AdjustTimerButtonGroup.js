import React from 'react';
import { Button } from 'react-materialize';
import { func } from 'prop-types';

const AdjustTimerButtonGroup = props => {
  return (
    <tr>
      <td>
        <div className="hours-set">
          <Button
            id="hours-increase"
            className="plus-btn blue"
            onClick={props.increaseHours}
          >
            +
          </Button>
          <Button
            id="hours-decrease"
            className="minus-btn blue"
            onClick={props.decreaseHours}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="minutes-set">
          <Button
            id="minutes-increase"
            className="plus-btn blue"
            onClick={props.increaseMinutes}
          >
            +
          </Button>
          <Button
            id="minutes-decrease"
            className="minus-btn blue"
            onClick={props.decreaseMinutes}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="seconds-set">
          <Button
            waves="light"
            id="seconds-increase"
            className="plus-btn blue"
            onClick={props.increaseSeconds}
          >
            +
          </Button>
          <Button
            id="seconds-decrease"
            className="minus-btn blue"
            onClick={props.decreaseSeconds}
          >
            -
          </Button>
        </div>
      </td>
    </tr>
  );
};

AdjustTimerButtonGroup.propTypes = {
  increaseHours: func.isRequired,
  decreaseHours: func.isRequired,
  increaseMinutes: func.isRequired,
  decreaseMinutes: func.isRequired,
  increaseSeconds: func.isRequired,
  decreaseSeconds: func.isRequired
};

export default AdjustTimerButtonGroup;
