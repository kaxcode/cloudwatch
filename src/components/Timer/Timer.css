.Timer__Heading,
.Timer__Hours,
.Timer__Minutes,
.Timer__Seconds,
.Timer__HoursCount,
.Timer__MinutesCount,
.Timer__SecondsCount {
  text-align: center;
}

.Timer__Card {
  min-width: 450px;
}

.Timer__HoursCount,
.Timer__MinutesCount,
.Timer__SecondsCount {
  margin: 0;
}

.Timer__StartBtn,
.Timer__PauseBtn,
.Timer__CleartBtn {
  margin-left: 8px;
}
