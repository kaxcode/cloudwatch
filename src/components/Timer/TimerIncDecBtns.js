import React from 'react';
import { Button } from 'react-materialize';
import { func, number, bool } from 'prop-types';

const Timer = props => {
  return (
    <tr>
      <td>
        <div className="hours-set">
          <Button
            id="hours-increase"
            className="plus-btn blue"
            onClick={props.increaseHours}
          >
            +
          </Button>
          <Button
            id="hours-decrease"
            className="minus-btn blue"
            onClick={props.decreaseHours}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="minutes-set">
          <Button
            id="minutes-increase"
            className="plus-btn blue"
            onClick={props.increaseMinutes}
          >
            +
          </Button>
          <Button
            id="minutes-decrease"
            className="minus-btn blue"
            onClick={props.decreaseMinutes}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="seconds-set">
          <Button
            waves="light"
            id="seconds-increase"
            className="plus-btn blue"
            onClick={props.increaseSeconds}
          >
            +
          </Button>
          <Button
            id="seconds-decrease"
            className="minus-btn blue"
            onClick={props.decreaseSeconds}
          >
            -
          </Button>
        </div>
      </td>
    </tr>
  );
};

Timer.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  increaseHours: func.isRequired,
  decreaseHours: func.isRequired,
  increaseMinutes: func.isRequired,
  decreaseMinutes: func.isRequired,
  increaseSeconds: func.isRequired,
  decreaseSeconds: func.isRequired,
  timeRemaining: number,
  startClicked: bool
};

export default Timer;
