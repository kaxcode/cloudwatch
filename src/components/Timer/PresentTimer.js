import React from 'react';
import {
  secondsToHour,
  secondsToMinutes,
  seconds
} from '../../utils/humanizeTimer';
import Alert from '../Alert/Alert';
import { func, number, bool, object } from 'prop-types';
import MessageBoard from '../MessageBoard/MessageBoard.js';
import Nav from '../Nav/Nav.js';
import './Timer.css';

class PresentTimer extends React.Component {
  componentDidMount() {
    this.forceUpdateInterval = setInterval(() => this.forceUpdate(), 50);
  }
  render() {
    return (
      <main className="Timer">
        <Alert
          onDismiss={this.props.handleDismiss}
          show={this.props.startClicked && this.props.timeRemaining <= 0}
        />
        <Nav />
        <div className="Timer__Container">
          <div className="Timer__Counter">
            <div className="Timer__HoursCount">
              {secondsToHour(localStorage.getItem('timeRemaining'))}
            </div>
            <div className="Timer__Delimiter">:</div>
            <div className="Timer__MinutesCount">
              {secondsToMinutes(localStorage.getItem('timeRemaining'))}
            </div>
            <div className="Timer__Delimiter">:</div>
            <div className="Timer__SecondsCount">
              {seconds(localStorage.getItem('timeRemaining'))}
            </div>
          </div>
          <MessageBoard
            location={this.props.location}
            timeRemaining={this.props.timeRemaining}
          />
        </div>
      </main>
    );
  }
}

PresentTimer.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  timeRemaining: number,
  startClicked: bool,
  location: object
};

export default PresentTimer;
