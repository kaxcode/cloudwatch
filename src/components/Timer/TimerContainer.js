import React from 'react';
import Timer from './Timer.js';

const HOURS = 3600;
const MINUTES = 60;
const SECONDS = 1;
const MAX_TIME = 24 * 60 * 60;
const MIN_TIME = 0;

class TimerContainer extends React.Component {
  state = {
    timeRemaining: 0,
    startClicked: false
  };

  timer = null;

  increaseHours = () => {
    if (this.state.timeRemaining >= MAX_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining + HOURS });
  };

  decreaseHours = () => {
    if (this.state.timeRemaining < HOURS) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining - HOURS });
  };

  increaseMinutes = () => {
    if (this.state.timeRemaining >= MAX_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining + MINUTES });
  };

  decreaseMinutes = () => {
    if (this.state.timeRemaining < MINUTES) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining - MINUTES });
  };

  increaseSeconds = () => {
    if (this.state.timeRemaining >= MAX_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining + SECONDS });
  };

  decreaseSeconds = () => {
    if (this.state.timeRemaining <= MIN_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining - SECONDS });
  };

  handleStart = () => {
    if (this.state.startClicked) {
      return;
    }
    clearInterval(this.timer);
    this.setState({ startClicked: true });
    this.timer = setInterval(this.tick, 1000);
  };

  handlePause = () => {
    clearInterval(this.timer);
    this.setState({ startClicked: false });
  };

  handleClear = () => {
    this.setState({
      startClicked: false,
      timeRemaining: 0
    });
  };

  tick = () => {
    if (this.state.timeRemaining <= 0) {
      return this.handleClear();
    }

    this.setState({
      timeRemaining: this.state.timeRemaining - 1
    });
  };

  handleDismiss = () => {
    this.setState({
      startClicked: false
    });
    clearInterval(this.timer);
  };

  render() {
    return (
      <Timer
        handleDismiss={this.handleDismiss}
        increaseHours={this.increaseHours}
        decreaseHours={this.decreaseHours}
        increaseMinutes={this.increaseMinutes}
        decreaseMinutes={this.decreaseMinutes}
        increaseSeconds={this.increaseSeconds}
        decreaseSeconds={this.decreaseSeconds}
        onStart={this.handleStart}
        onPause={this.handlePause}
        onClear={this.handleClear}
        startClicked={this.state.startClicked}
        timeRemaining={this.state.timeRemaining}
      />
    );
  }
}

export default TimerContainer;
