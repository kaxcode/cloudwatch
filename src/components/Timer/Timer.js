import React from 'react';
import { Row, Button, Card, Col, Table } from 'react-materialize';
import {
  secondsToHour,
  secondsToMinutes,
  seconds
} from '../../utils/humanizeTimer';
import Alert from '../Alert/Alert';
import { func, number, bool } from 'prop-types';
import TimerIncDecBtns from './TimerIncDecBtns';

const Timer = props => {
  return (
    <Row>
      <Alert
        msg="Time is up!"
        onDismiss={props.handleDismiss}
        show={props.startClicked && props.timeRemaining <= 0}
      />
      <Col s={12} m={3} offset="m2 l4">
        <Card className="white parent-container" textClassName="black-text">
          <h3>Timer</h3>
          <Table className="setters">
            <thead>
              <tr>
                <th className="timer-heading">Hours</th>
                <th className="timer-heading">Minutes</th>
                <th className="timer-heading">Seconds</th>
              </tr>
            </thead>
            <tbody>
              <TimerIncDecBtns
                increaseHours={props.increaseHours}
                decreaseHours={props.decreaseHours}
                increaseMinutes={props.increaseMinutes}
                decreaseMinutes={props.decreaseMinutes}
                increaseSeconds={props.increaseSeconds}
                decreaseSeconds={props.decreaseSeconds}
              />
              <tr>
                <td>
                  <h1 className="timer-count">
                    {secondsToHour(props.timeRemaining)}
                  </h1>
                </td>
                <td>
                  <h1 className="timer-count">
                    {secondsToMinutes(props.timeRemaining)}
                  </h1>
                </td>
                <td>
                  <h1 className="timer-count">
                    {seconds(props.timeRemaining)}
                  </h1>
                </td>
              </tr>
              <tr>
                <td>
                  <Button
                    waves="light"
                    className="green start-btn"
                    onClick={props.onStart}
                  >
                    Start
                  </Button>
                </td>
                <td>
                  <Button
                    waves="light"
                    className="red pause-btn"
                    onClick={props.onPause}
                  >
                    Pause
                  </Button>
                </td>
                <td>
                  <Button
                    waves="light"
                    className="yellow darken-3 clear-btn"
                    onClick={props.onClear}
                  >
                    Clear
                  </Button>
                </td>
              </tr>
            </tbody>
          </Table>
        </Card>
      </Col>
    </Row>
  );
};

Timer.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  increaseHours: func.isRequired,
  decreaseHours: func.isRequired,
  increaseMinutes: func.isRequired,
  decreaseMinutes: func.isRequired,
  increaseSeconds: func.isRequired,
  decreaseSeconds: func.isRequired,
  timeRemaining: number,
  startClicked: bool
};

export default Timer;
