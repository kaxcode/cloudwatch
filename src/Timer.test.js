import React from 'react';
import Timer from './Timer';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders correctly', () => {
  const tree = renderer.create(<Timer />).toJSON();
  expect(tree).toMatchSnapshot();
});

jest.useFakeTimers();

test('Timer runs after click', () => {
  const timer = shallow(<Timer />);
  timer.instance().startTimer();

  expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
});
