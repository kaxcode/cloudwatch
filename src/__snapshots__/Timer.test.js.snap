// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="col s12 m3 offset-m2 offset-l4"
  >
    <div
      className="white parent-container card"
    >
      <div
        className="card-content black-text"
      >
        <div>
          <h3>
            CloudWatch
          </h3>
          <h3>
            00:00:00
          </h3>
          <div>
            <button
              className="btn waves-effect waves-light green"
              disabled={false}
              onClick={[Function]}
            >
              Start
            </button>
            <button
              className="btn waves-effect waves-light red"
              disabled={false}
              onClick={[Function]}
            >
              Pause
            </button>
            <button
              className="btn waves-effect waves-light yellow darken-3"
              disabled={false}
              onClick={[Function]}
            >
              Clear
            </button>
          </div>
        </div>
      </div>
      <div
        className="card-reveal"
      >
        <span
          className="card-title grey-text text-darken-4"
        >
          <i
            className="material-icons right"
          >
            close
          </i>
        </span>
      </div>
    </div>
  </div>
</div>
`;
